fish-rust/src/wutil/wrealpath.rs

```diff
@@ -0,0 +1,72 @@
+use std::{ffi::OsStr, fs::canonicalize, os::unix::prelude::OsStrExt};
+
+use cxx::{CxxString, UniquePtr};
+
+use crate::{
+    ffi::{make_string, str2wcstring, wcs2string},
+    wchar::{wstr, WString},
+    wchar_ffi::{WCharFromFFI, WCharToFFI},
+};
+
+pub fn wrealpath(pathname: &wstr) -> Option<WString> {
+    if pathname.is_empty() {
+        return None;
+    }
+
+    let mut real_path: UniquePtr<CxxString> = make_string("");
+    let narrow_path_ptr: UniquePtr<CxxString> = wcs2string(&pathname.to_ffi());
+    // We can't `pop` from a `UniquePtr<CxxString>` so we copy it into a vector
+    let mut narrow_path = narrow_path_ptr.as_bytes().to_vec();
```

Comment:
please use native Rust strings wherever possible. Put the conversions close to the FFI boundary.
For example

	let narrow_path = wcs2string(&pathname.to_ffi()).from_ffi();

or is there a reason we can't use WString like embedded nul characters?

Reply:
>or is there a reason we can't use WString like embedded nul characters?

Paths can't have nul characters.

Reply:
I originally used the FFI strings because it was easier to treat them as a "bag of bytes" which is how Rust handles inputs to `std::fs::canonicalize`. I can see why it might be better to minimize the use of FFI types

Reply:
We also don't have an implementation of `WCharFromFFI<WString>` for `cxx::UniquePtr<cxx::CxxString>`

Reply:
Converting "string" to "wstring" is what str2wcstring does.

I can't see any place we use it on the rust side yet, so this might be the first.

I don't believe we want a WcharFromFFI for this, because it's a more involved operation?

Reply:
The original implementation of `wrealpath` does the opposite here, converting from a wide string to a "regular" string (i.e. basically an array of bytes) via `wcs2string` so it can be passed to the `realpath` function. If we want to keep the original behavior while avoiding FFI types shouldn't we use a `Vec<u8>` or maybe an `OsStr` here?

Reply:
Upon further experimentation it seems like `Vec<u8>` is the most suitable to replicate the structure of the original code. I'll push up a change for y'all to check out.